tests/network/test_continue_request.py

```diff
@@ -0,0 +1,241 @@
+    } == exception_info.value.args[0]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("phase", [
+    "responseStarted",
+    pytest.param("authRequired",
+                 marks=pytest.mark.xfail(
+                     reason='TODO: #644: implement "authRequired" phase'))
+])
+async def test_continue_request_invalid_phase(websocket, context_id, phase):
+    # TODO: make offline.
+    url = "https://www.example.com/"
+
+    network_id = await create_dummy_blocked_request(websocket,
+                                                    context_id,
+                                                    url=url,
+                                                    phases=[phase])
+
+    with pytest.raises(Exception) as exception_info:
+        await execute_command(
+            websocket, {
+                "method": "network.continueRequest",
+                "params": {
+                    "request": network_id,
+                    "url": url,
+                },
+            })
+    assert {
+        "error": "invalid argument",
+        "message": f"Blocked request for network id '{network_id}' is not in 'BeforeRequestSent' phase"
+    } == exception_info.value.args[0]
+
+
+@pytest.mark.asyncio
+async def test_continue_request_invalid_url(websocket, context_id):
+    # TODO: make offline.
+    url = "https://www.example.com/"
+    invalid_url = '%invalid%'
+
+    network_id = await create_dummy_blocked_request(
+        websocket, context_id, url=url, phases=["beforeRequestSent"])
+
+    with pytest.raises(Exception) as exception_info:
+        await execute_command(
+            websocket, {
+                "method": "network.continueRequest",
+                "params": {
+                    "request": network_id,
+                    "url": invalid_url,
+                },
+            })
+    assert {
+        "error": "invalid argument",
+        "message": f"Invalid URL '{invalid_url}': TypeError: Failed to construct 'URL': Invalid URL",
+    } == exception_info.value.args[0]
+
+
+@pytest.mark.asyncio
+async def test_continue_request_completes(websocket, context_id):
+    # TODO: make offline.
+    url = "https://www.example.com/"
+
+    await subscribe(websocket, ["cdp.Fetch.requestPaused"])
+
+    await execute_command(
+        websocket, {
+            "method": "network.addIntercept",
+            "params": {
+                "phases": ["beforeRequestSent"],
+                "urlPatterns": [{
+                    "type": "string",
+                    "pattern": url,
+                }, ],
+            },
+        })
+    await send_JSON_command(
+        websocket, {
+            "method": "browsingContext.navigate",
+            "params": {
+                "url": url,
+                "context": context_id,
+            }
+        })
+    event_response = await wait_for_event(websocket, "cdp.Fetch.requestPaused")
+    network_id = event_response["params"]["params"]["networkId"]
```

Comment:
This looks odd. Do we really need waiting for the CDP request first?

Reply:
The responseStarted event is never emitted in this case, that's why.

I am not yet sure whether we should manually emit it (then we could use it in the test).

Either way: For now, this is out of scope of this PR (let's keep using CDP events for consistency, then we can revisit all tests later and change them to BiDi events if warranted).

But let me know if you think we should emit the events manually.

Reply:
Agree to move it out of scope, but eventually we should get rid of this CDP hack.

Reply:
Actually, I don't quite follow the e2e test. If there is no BiDi event pointing the request is blocked, how user is expected to continue it?

Reply:
Right, there is `responseStarted ` event, as irequest is intercepted on the request state. Butthere should be an event `network.beforeRequestSent` event with `isBlocked: true` should be emitted. Why cannot we use it?

Reply:
> But there should be an event network.beforeRequestSent event with isBlocked: true should be emitted. Why cannot we use it?

So basically you're answering 'yes' to the question I made above. OK. Let's emit those events in a separate PR then. We will need to revisit all methods to do so.

Reply:
Resolving but made a note of it.

Reply:
https://github.com/GoogleChromeLabs/chromium-bidi/pull/1400